### ISSUE TITLE
Prediction from Model

### ISSUE THREAD

**Issue by muschellij2** [skeydan/lesion_segmentation#3]
In FLEXCONN `FLEXCONN_Test.py`, we have:

```
def apply_model(image1, image2, pred_model):

    dim = image1.shape
    dim_2d = (1, dim[0], dim[1], 1)

    slice_2d_t1 = np.zeros(dim_2d, dtype=np.float16)
    slice_2d_fl = np.zeros(dim_2d, dtype=np.float16)
    output_image = np.zeros(dim, dtype=np.float16)

    for k in range(dim[2]):
        slice_2d_t1[0, :, :, 0] = image1[:, :, k]
        slice_2d_fl[0, :, :, 0] = image2[:, :, k]
        pred = pred_model.predict([slice_2d_t1, slice_2d_fl])
        output_image[:, :, k] = pred[0, :, :, 0]

    return output_image
```

It seems as though the model is predicting on a slice of an image, but not on patches of the image.  is this how you understand the code as well?  It seems odd.

I assumed the process was:
1.  Take image, make patches.  Use whole brain mask or another mask of "patches"
2.  Predict on a patch-based model.  
3.  Take prediction from center voxel of patch as the probability of lesion for that voxel.
4.  Aggregating all the patches above make a full prediction image.

Is that how you see the process?

**Comment by muschellij2** [comment]
Instead of 3., you could also take all the patches that have voxel `v` and then average over these predictions

**Comment by skeydan** [comment]
yeah, hadn't looked at the test code yet, but now having looked, yes it seems they're doing it on whole images now (whole in 2d that is)...
Thinking about it, you get a probability per pixel (voxel) either way, right?

Does this mean the whole patches thing is "just" to arrange the training data in a manageable way?
I shouldn't say "just", because clearly if I can hardly train this with batch size 1 and patch size 35, whole images should be pretty unfeasible even for the big GPUs...

BTW technically this works bc they define the input layer like so
```
t1 = Input((None, None, 1))
```

**Comment by muschellij2** [comment]
Ahh - I see that this "works", but I'm still unsure if that gives you the
right output?  I wonder if you get the same probability from passing in the
slice versus the patch?

I think the patches will dictate the number of parameters in the models and
changing patchsize would change parameters, correct?  So I don't think it's
just arrangement.



John

**Comment by skeydan** [comment]
Number of parameters should depend on filter sizes (and counts) only  as the network is fully convolutional...

But should you get the same output? Not sure, I guess one  would have to calculate the receptive fields after all conv layers and see if that's above the common patch sizes they're mentioning...
Pretty likely you get (slightly) different results I'd say but then, question is does it matter when in the end all you do is thresholding...?
